tests/Aql/InsertExecutorTest.cpp

```diff
@@ -0,0 +1,428 @@
+  std::vector<size_t> param = GetParam();
+  for (auto i : param) {
+    std::string query = std::string("FOR i IN 1..") + std::to_string(i) +
+                        " INSERT { value: i } INTO " + collectionName;
+
+    AssertQueryHasResult(vocbase, query, VPackSlice::emptyArraySlice());
+
+    for (size_t j = 1; j <= i; j++) {
+      insertedVals.push_back(j);
+    }
+  }
+
+  std::sort(std::begin(insertedVals), std::end(insertedVals));
+  VPackBuilder builder;
+  builder.openArray();
+  for (auto i : insertedVals) {
+    builder.add(VPackValue(i));
+  }
+  builder.close();
+  AssertQueryHasResult(vocbase, checkQuery, builder.slice());
+}
+
+TEST_P(InsertExecutorTestCounts, insert_multiple_with_return) {
+  std::vector<size_t> insertedVals = {};
+  auto const bindParameters = VPackParser::fromJson("{ }");
+
+  std::vector<size_t> param = GetParam();
+
+  for (auto i : param) {
+    std::string query = std::string("FOR i IN 1..") + std::to_string(i) +
+                        " INSERT { value: i } INTO " + collectionName +
+                        " RETURN NEW ";
+    auto result = arangodb::tests::executeQuery(vocbase, query, bindParameters);
+    ASSERT_TRUE(result.ok());
+    for (size_t j = 1; j <= i; j++) {
+      insertedVals.push_back(j);
+    }
+  }
+
+  std::sort(std::begin(insertedVals), std::end(insertedVals));
+  VPackBuilder builder;
+  builder.openArray();
+  for (auto i : insertedVals) {
+    builder.add(VPackValue(i));
+  }
+  builder.close();
+  AssertQueryHasResult(vocbase, checkQuery, builder.slice());
+}
+
+INSTANTIATE_TEST_CASE_P(
+    InsertExecutorTestInstance, InsertExecutorTestCounts,
+    testing::Values(std::vector<size_t>{1}, std::vector<size_t>{100},
+                    std::vector<size_t>{999}, std::vector<size_t>{1000},
+                    std::vector<size_t>{1001}, std::vector<size_t>{1, 100, 1000, 1000, 900},
+                    std::vector<size_t>{10, 10, 10, 10, 10, 100, 100, 10, 100,
+                                        1000, 1000, 900, 10, 100}));
+
+// OLD is a keyword, but only sometimes. In particular in insert queries it
+// isn't unless overwrite: true. Yes, really.
+TEST_F(InsertExecutorTest, insert_return_old) {
+  std::string query = std::string("FOR i IN 1..1 INSERT { value: i } INTO ") +
+                      collectionName + " RETURN OLD";
+
+  AssertQueryFailsWith(vocbase, query, TRI_ERROR_ARANGO_DATA_SOURCE_NOT_FOUND);
+}
+
+TEST_P(InsertExecutorTestCount, insert_with_key) {
+  std::string query =
+      std::string("FOR i IN 1.." + nDocsString +
+                  " INSERT { _key: TO_STRING(i), value: i } INTO ") +
+      collectionName + " SORT NEW.value RETURN NEW.value";
+
+  VPackBuilder builder;
+  builder.openArray();
+  for (size_t i = 1; i <= nDocs; i++) {
+    builder.add(VPackValue(i));
+  }
+  builder.close();
+
+  AssertQueryHasResult(vocbase, query, builder.slice());
+}
+
+TEST_P(InsertExecutorTestCount, insert_with_key_and_overwrite) {
```

Comment:
This test is correct and good, however I wonder what is the difference between INSERT overwrite:true
and REPSERT?

Reply:
with `REPSERT` you can insert a different document for the `INSERT` case vs the `REPLACE` case, I guess.

Reply:
with `REPSERT` you can insert a different document for the `INSERT` case vs the `REPLACE` case, I guess.